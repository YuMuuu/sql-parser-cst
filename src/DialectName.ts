export type DialectName = "sqlite" | "mysql";
