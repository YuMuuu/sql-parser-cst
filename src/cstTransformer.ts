import { Node } from "./cst/Node";

// Types are getting somewhat intimidating in this file.
// Learned this TypeScript-Fu from here:
// https://stackoverflow.com/questions/64092736/alternative-to-switch-statement-for-typescript-discriminated-union

/**
 * A map with a transform function for each Node type, like:
 *
 *     { select_clause: (node: SelectClause) => {},
 *       from_clause: (node: FromClause) => {},
 *       paren_expr: (node: ParenExpr) => {},
 *       ... }
 */
export type FullTransformMap<T> = {
  [K in Node["type"]]: (node: Extract<Node, { type: K }>) => T;
};

/**
 * Creates a function that transforms the whole syntax tree to data type T.
 * @param map An object with a transform function for each CST node type
 */
export function cstTransformer<T>(
  map: Partial<FullTransformMap<T>>
): (node: Node) => T {
  return (node: Node) => {
    const fn = map[node.type] as (
      e: Extract<Node, { type: typeof node["type"] }>
    ) => T;
    if (!fn) {
      if (!node.type) {
        throw new Error(`No type field on node: ${JSON.stringify(node)}`);
      }
      throw new Error(`No transform map entry for ${node.type}`);
    }
    return fn(node);
  };
}
