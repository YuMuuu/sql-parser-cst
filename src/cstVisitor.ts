import { Node } from "./sql";

/**
 * A map with a visitor function for each Node type, like:
 *
 *     { select_clause: (node: SelectClause) => {},
 *       from_clause: (node: FromClause) => {},
 *       paren_expr: (node: ParenExpr) => {},
 *       ... }
 */
export type FullVisitorMap = {
  [K in Node["type"]]: (node: Extract<Node, { type: K }>) => void;
};

/**
 * Creates a function that visits the whole syntax tree.
 * @param map An object with a visitor functions
 */
export function cstVisitor(map: Partial<FullVisitorMap>): (node: Node) => void {
  const visit = (node: Node) => {
    const visitType = map[node.type] as (
      e: Extract<Node, { type: typeof node["type"] }>
    ) => void;

    // Visit the node itself
    if (visitType) {
      visitType(node);
    }

    // Visit all children
    for (const child of Object.values(node)) {
      if (isNode(child)) {
        visit(child);
      } else if (child instanceof Array) {
        child.filter(isNode).forEach((childNode) => visit(childNode));
      }
    }
  };
  return visit;
}

const isNode = (value: any): value is Node =>
  typeof value === "object" && typeof value.type === "string";
