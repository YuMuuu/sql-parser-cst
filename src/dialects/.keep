This file exists to ensure the dir is included to git
