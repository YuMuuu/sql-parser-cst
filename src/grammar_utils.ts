/** Identity function */
export const identity = (x: any) => x;

/** Last item in array */
const last = (arr: any) => arr[arr.length - 1];

/** Creates new array with first item replaced by value */
const setFirst = ([oldFirst, ...rest]: any, value: any) => {
  return [value, ...rest];
};

/** Creates new array with last item replaced by value */
const setLast = (array: any, value: any) => {
  const rest = array.slice(0, -1);
  return [...rest, value];
};

/** Attaches optional leading whitespace to AST node, or to array of AST nodes (to the first in array) */
export const leading = (node: any, ws: any): any => {
  if (node instanceof Array) {
    // Add leading whitespace to first item in array
    return setFirst(node, leading(node[0], ws));
  }
  if (typeof node !== "object") {
    throw new Error(
      `Expected Node object, instead got ${JSON.stringify(node)}`
    );
  }
  if (ws && ws.length) {
    if (node.leading) {
      throw new Error("leading(): Node already has leading whitespace");
    }
    return { ...node, leading: ws };
  }
  return node;
};

/** Attaches optional trailing whitespace to AST node, or to array of AST nodes (to the last in array) */
export const trailing = (node: any, ws: any): any => {
  if (node instanceof Array) {
    // Add trailing whitespace to last item in array
    return setLast(node, trailing(last(node), ws));
  }
  if (typeof node !== "object") {
    throw new Error(
      `Expected Node object, instead got ${JSON.stringify(node)}`
    );
  }
  if (ws && ws.length) {
    if (node.trailing) {
      throw new Error("trailing(): Node already has trailing whitespace");
    }
    return { ...node, trailing: ws };
  }
  return node;
};

// Shorthand for attaching both trailing or leading whitespace
export const surrounding = (leadingWs: any, node: any, trailingWs: any) =>
  trailing(leading(node, leadingWs), trailingWs);

const deriveLoc = (binExpr: any) => {
  if (!binExpr.left.range) {
    return binExpr;
  }
  const start = binExpr.left.range[0];
  const end = binExpr.right.range[1];
  return { ...binExpr, range: [start, end] };
};

export function createBinaryExprChain(
  head: any,
  tail: any,
  type: any = "binary_expr"
) {
  return tail.reduce(
    (left: any, [c1, op, c2, right]: any[]) =>
      deriveLoc(createBinaryExpr(left, c1, op, c2, right, type)),
    head
  );
}

export function createBinaryExpr(
  left: any,
  c1: any,
  op: any,
  c2: any,
  right: any,
  type: any = "binary_expr"
) {
  return {
    type,
    operator: op,
    left: trailing(left, c1),
    right: leading(right, c2),
  };
}

const deriveJoinLoc = (join: any) => {
  if (!join.left.range) {
    return join;
  }
  const start = join.left.range[0];
  const end = (join.spec || join.right).range[1];
  return { ...join, range: [start, end] };
};

export function createJoinExprChain(head: any, tail: any) {
  return tail.reduce(
    (left: any, [c1, op, c2, right, spec]: any[]) =>
      deriveJoinLoc(createJoinExpr(left, c1, op, c2, right, spec)),
    head
  );
}

function createJoinExpr(
  left: any,
  c1: any,
  op: any,
  c2: any,
  right: any,
  spec: any
) {
  return {
    type: "join_expr",
    left: trailing(left, c1),
    operator: op,
    right: leading(right, c2),
    specification: read(spec),
  };
}

export function createPrefixOpExpr(op: any, expr: any) {
  return {
    type: "prefix_op_expr",
    operator: op,
    expr: expr,
  };
}

export function createPostfixOpExpr(op: any, expr: any) {
  return {
    type: "postfix_op_expr",
    expr: expr,
    operator: op,
  };
}

export const createKeyword = (text: any) => ({ type: "keyword", text });

const whitespaceType: Record<string, boolean> = {
  space: true,
  newline: true,
  line_comment: true,
  block_comment: true,
};

// True when dealing with whitespace array (as returned by __ rule)
const isWhitespace = (item: any) => {
  if (!(item instanceof Array)) {
    return false;
  }
  if (item.length === 0) {
    return true;
  }
  return Boolean(whitespaceType[item[0].type as string]);
};

/**
 * Given array of syntax nodes and whitespace or single node or null,
 * associates whitespace with syntax nodes.
 *
 * @param {(Node | Whitespace)[] | Node | null} items
 * @return {Node[] | Node | undefined}
 */
export const read = (items: any) => {
  if (!items) {
    return undefined; // convert null to undefined
  }
  if (!(items instanceof Array)) {
    return items; // leave single syntax node as-is
  }

  // associate leading/trailing whitespace with nodes
  const nodes: any[] = [];
  let leadingWhitespace = undefined;
  for (const it of items) {
    if (isWhitespace(it)) {
      if (nodes.length > 0) {
        nodes[nodes.length - 1] = trailing(nodes[nodes.length - 1], it);
      } else {
        leadingWhitespace = it;
      }
    } else {
      if (leadingWhitespace) {
        nodes.push(leading(it, leadingWhitespace));
        leadingWhitespace = undefined;
      } else {
        nodes.push(it);
      }
    }
  }
  return nodes.length > 1 ? nodes : nodes[0];
};

export const readCommaSepList = (head: any, tail: any) => {
  const items = [head];
  for (const [c1, comma, c2, expr] of tail) {
    const lastIdx = items.length - 1;
    items[lastIdx] = trailing(items[lastIdx], c1);
    items.push(leading(expr, c2));
  }
  return items;
};

export const readSpaceSepList = (head: any, tail: any) => {
  const items = [head];
  for (const [c, expr] of tail) {
    items.push(leading(expr, c));
  }
  return items;
};

export const createIdentifier = (text: any) => ({ type: "identifier", text });

export const createAlias = (expr: any, _alias: any) => {
  if (!_alias) {
    return expr;
  }
  const [c, partialAlias] = _alias;
  return {
    type: "alias",
    expr: trailing(expr, c),
    ...partialAlias,
  };
};

export const createParenExpr = (c1: any, expr: any, c2: any) => {
  return {
    type: "paren_expr",
    expr: surrounding(c1, expr, c2),
  };
};

export const createExprList = (head: any, tail: any) => {
  return {
    type: "expr_list",
    items: readCommaSepList(head, tail),
  };
};

export const hasParamType = (name: any, options: any) => {
  return options?.paramTypes?.includes(name);
};

const isEnabledWhitespace = (ws: any, options: any) =>
  (options?.preserveComments &&
    (ws.type === "line_comment" || ws.type === "block_comment")) ||
  (options?.preserveNewlines && ws.type === "newline") ||
  (options?.preserveSpaces && ws.type === "space");

export const filterEnabledWhitespace = (items: any, options: any) => {
  return items.filter((ws: any) => isEnabledWhitespace(ws, options));
};

let getRange: Function;

export const setRangeFunction = (fn: Function) => {
  getRange = fn;
};

let getOptions: Function;

export const setOptionsFunction = (fn: Function) => {
  getOptions = fn;
};

export const loc = (node: any) => {
  if (!getOptions()?.includeRange) {
    return node;
  }
  const { start, end } = getRange();
  return { ...node, range: [start, end] };
};
