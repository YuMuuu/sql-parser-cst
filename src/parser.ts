import { parse as mysql, Ast } from "../pegjs/mysql";
import { parse as sqlite } from "../pegjs/sqlite";
import { show as showSql } from "./show";

export function parse(sql: string, dialect: "mysql" | "sqlite" = "mysql"): Ast {
  switch (dialect) {
    case "mysql":
      return mysql(sql);
    case "sqlite":
      return sqlite(sql);
  }
}

export function show(ast: Ast): string {
  return showSql(ast, "; ").trimEnd();
}
