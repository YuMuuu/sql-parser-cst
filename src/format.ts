import { Node, Program } from "../pegjs/sql";
import { cstTransformer } from "./cstTransformer";

export type Line = {
  layout: "line";
  indent?: number;
  items: Layout[];
};

export type Layout = Line | string | Layout[];

export function format(node: Program) {
  const lines = unroll(layout(node));
  if (!(lines instanceof Array) || !lines.every(isLine)) {
    throw new Error(`Expected array of lines, instead got ${lines}`);
  }
  return serialize(lines);
}

export function serialize(lines: Line[]): string {
  const INDENT = "  ";
  return lines
    .map((line) => INDENT.repeat(line.indent || 0) + line.items.join(""))
    .join("\n");
}

export const line = (...items: Layout[]): Line => ({ layout: "line", items });
export const indent = (...items: Layout[]): Line => ({
  layout: "line",
  indent: 1,
  items,
});

const isLine = (item: Layout): item is Line =>
  typeof item === "object" && (item as any).layout === "line";

export function unroll(item: Layout): Layout {
  if (isLine(item)) {
    return unrollLine(item);
  }
  if (item instanceof Array) {
    return unrollArray(item);
  }
  return item;
}

function unrollArray(arr: Layout[]): Layout[] {
  return arr.flatMap(unroll);
}

function unrollLine(line: Line): Line[] {
  const lineItems = unrollArray(line.items);
  if (lineItems.every(isLine)) {
    return lineItems.map((subLine) => {
      if (line.indent) {
        return { ...subLine, indent: line.indent + (subLine.indent || 0) };
      } else {
        return subLine;
      }
    });
  }
  return [{ ...line, items: lineItems }];
}

export function layout(node: Node): Layout {
  return layoutNode(node);
}

const layoutNode = cstTransformer<Layout>({
  program: (node) => node.statements.map(layout),
  empty_statement: () => "",

  // SELECT statement
  compound_select_statement: (node) => "",
  select_statement: (node) => node.clauses.map(layout),
  // WITH
  with_clause: (node) => "",
  common_table_expression: (node) => "",
  // SELECT
  select_clause: (node) => [
    line(layout(node.selectKw)),
    indent(
      ...node.columns
        .map(layout)
        .map((col, i) =>
          i < node.columns.length - 1 ? line(col, ",") : line(col)
        )
    ),
  ],
  // FROM
  from_clause: (node) => "",
  join: (node) => "",
  join_on_specification: (node) => "",
  join_using_specification: (node) => "",
  sort_specification: (node) => "",
  // WHERE .. GROUP BY .. HAVING .. ORDER BY .. PARTITION BY
  where_clause: (node) => "",
  group_by_clause: (node) => "",
  having_clause: (node) => "",
  order_by_clause: (node) => "",
  partition_by_clause: (node) => "",
  // WINDOW
  window_clause: (node) => "",
  named_window: (node) => "",
  window_definition: (node) => "",
  // LIMIT
  limit_clause: (node) => "",
  // VALUES
  values_clause: (node) => "",

  // Window frame
  frame_clause: (node) => "",
  frame_between: (node) => "",
  frame_bound_current_row: (node) => "",
  frame_bound_preceding: (node) => "",
  frame_bound_following: (node) => "",
  frame_unbounded: (node) => "",
  frame_exclusion: (node) => "",

  // CREATE TABLE statement
  create_table_statement: (node) => "",
  column_definition: (node) => "",
  column_option_nullable: (node) => "",
  column_option_auto_increment: (node) => "",
  column_option_key: (node) => "",
  column_option_default: (node) => "",
  column_option_comment: (node) => "",

  // DROP TABLE statement
  drop_table_statement: (node) => "",

  // INSERT INTO statement
  insert_statement: (node) => "",
  insert_option: (node) => "",
  default_values: (node) => "",
  default: (node) => "",

  // Expressions
  expr_list: (node) => "",
  paren_expr: (node) => "",
  binary_expr: (node) => "",
  unary_expr: (node) => "",
  func_call: (node) => "",
  distinct_arg: (node) => "",
  cast_expr: (node) => "",
  cast_arg: (node) => "",
  over_arg: (node) => "",
  between_expr: (node) => "",
  datetime: (node) => "",
  string_with_charset: (node) => "",
  case_expr: (node) => "",
  case_when: (node) => "",
  case_else: (node) => "",
  interval_expr: (node) => "",

  // Data types
  data_type: (node) => "",

  // Tables & columns
  column_ref: (node) => "",
  table_ref: (node) => "",
  alias: (node) => "",
  all_columns: () => "*",

  // Basic language elements
  keyword: (node) => node.text,
  identifier: (node) => node.text,
  string: (node) => node.text,
  number: (node) => node.text,
  bool: (node) => node.text,
  null: (node) => node.text,
});
