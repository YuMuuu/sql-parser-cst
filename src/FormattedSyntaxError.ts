import { SyntaxError } from "./parser";

export class FormattedSyntaxError extends Error {
  constructor(e: SyntaxError, sql: string, fileName: string) {
    super(formatError(e, sql, fileName));
  }
}

function formatError(e: SyntaxError, sql: string, fileName: string): string {
  const lineNr = e.location.start.line;
  const colNr = e.location.start.column;
  const line = sql.split("\n")[lineNr - 1];
  const indent = "".padStart(String(lineNr).length);
  return `Syntax Error: ${e.message}
--> ${fileName}:${lineNr}:${colNr}
${indent} |
${lineNr} | ${line}
${indent} | ${"^".padStart(colNr)}`;
}
