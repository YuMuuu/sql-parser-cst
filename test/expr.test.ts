import { parse } from "../src/parser";

describe("expr", () => {
  function parseExpr(expr: string) {
    return parse(`SELECT ${expr}`).columns[0];
  }

  describe("operators", () => {
    // punctuation-based operators
    [
      "+",
      "-",
      "~",
      "*",
      "/",
      "%",
      "&",
      ">>",
      "<<",
      "^",
      "|",
      "~",
      ">=",
      ">",
      "<=",
      "<>",
      "<",
      "=",
      "!=",
    ].forEach((op) => {
      it(`parses ${op} operator`, () => {
        expect(parseExpr(`5 ${op} 7`)).toMatchSnapshot();
      });
    });

    it("parses chain of addition and subtraction", () => {
      expect(parseExpr(`5 + 6 - 8`)).toMatchSnapshot();
    });

    it("parses chain of multiplication and division", () => {
      expect(parseExpr(`2 * 7 / 3`)).toMatchSnapshot();
    });

    it("treats multiplication with higher precedence than addition", () => {
      expect(parseExpr(`6 + 7 * 3`)).toMatchSnapshot();
    });

    it("parses DIV operator", () => {
      expect(parseExpr(`8 DIV 4`)).toMatchSnapshot();
    });

    it("recognizes lowercase DIV operator", () => {
      expect(parseExpr(`8 div 4`)).toMatchSnapshot();
    });

    it("parses addition with comments", () => {
      expect(parseExpr(`6 /* com1 */ + /* com2 */ 7`)).toMatchSnapshot();
    });

    it("parses multiplication with comments", () => {
      expect(parseExpr(`6 /* com1 */ * /* com2 */ 7`)).toMatchSnapshot();
    });

    it("parses comparison with comments", () => {
      expect(parseExpr(`6 /* com1 */ < /* com2 */ 7`)).toMatchSnapshot();
    });

    it("parses IS operator", () => {
      expect(parseExpr(`7 IS 5`)).toMatchSnapshot();
    });

    it("parses IS NOT operator", () => {
      expect(parseExpr(`7 IS NOT 5`)).toMatchSnapshot();
    });

    it("parses IS NOT operator with comment inside", () => {
      expect(parseExpr(`7 IS /* com */ NOT 5`)).toMatchSnapshot();
    });
  });
});
