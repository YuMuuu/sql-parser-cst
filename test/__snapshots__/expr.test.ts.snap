// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`expr literal bit string 1`] = `
{
  "prefix": null,
  "text": "b'011001'",
  "type": "string",
}
`;

exports[`expr literal bit string 2`] = `
{
  "prefix": "_binary",
  "text": "b'011001'",
  "type": "string",
}
`;

exports[`expr literal boolean 1`] = `
{
  "text": "true",
  "type": "bool",
}
`;

exports[`expr literal boolean 2`] = `
{
  "text": "TRUE",
  "type": "bool",
}
`;

exports[`expr literal boolean 3`] = `
{
  "text": "false",
  "type": "bool",
}
`;

exports[`expr literal boolean 4`] = `
{
  "text": "FALSE",
  "type": "bool",
}
`;

exports[`expr literal double-quoted string 1`] = `
{
  "text": ""hello"",
  "type": "string",
}
`;

exports[`expr literal hex literal 1`] = `
{
  "prefix": null,
  "text": "0xAAFF11",
  "type": "string",
}
`;

exports[`expr literal hex literal 2`] = `
{
  "prefix": "_binary",
  "text": "0xAAFF11",
  "type": "string",
}
`;

exports[`expr literal hex string 1`] = `
{
  "prefix": null,
  "text": "x'AFC123'",
  "type": "string",
}
`;

exports[`expr literal hex string 2`] = `
{
  "prefix": "_binary",
  "text": "x'AFC123'",
  "type": "string",
}
`;

exports[`expr literal single-quoted string 1`] = `
{
  "text": "'hello'",
  "type": "string",
}
`;
